smolyay/basis.py

```diff
@@ -403,29 +482,393 @@ def _derivative(self, x):
+        """Evaluate all the derivative of basis functions in the set
+
+        Calls the derivative for all the basis function(s) and
+        evaluates at X.
+
+        Parameters
+        ----------
+        X : array-like
+            the points to evaluate.
+
+        X_domain : numpy array of shape (2,)
+            the lower and upper bounds of X.
+
+        n : int, optional
+            order of derivative. Default is 1.
+
+        Returns
+        -------
+        scalar or ndarray
+            the values of the basis functions.
+        """
+        new_X = self._scale_to_domain(numpy.asarray(X), X_domain)
+        if any(bf._is_complex for bf in self):
+            y = numpy.zeros([len(self)] + list(new_X.shape), dtype="complex_")
+        else:
+            y = numpy.zeros([len(self)] + list(new_X.shape))
+        for i in range(len(self)):
+            y[i, :] = self[i].derivative(new_X, n)
+        y *= ((self.domain[1] - self.domain[0]) / (X_domain[1] - X_domain[0])) ** n
+        return y
+
+
+class CustomBasisFunctionSet(BasisFunctionSet):
+    """Set of basis functions
 
     Parameters
     ----------
     basis_functions : list
         Basis functions in set.
 
+    Raises
+    ------
+    IndexError
+        Must have at least one basis function
+    ValueError
+        Basis functions must have the same domain.
     """
 
     def __init__(self, basis_functions):
+        if len(basis_functions) == 0:
+            raise IndexError("Must have at least one basis function.")
+        domain = basis_functions[0].domain
+        if any(not numpy.array_equal(domain, b.domain) for b in basis_functions[1:]):
+            raise ValueError("Basis functions must have the same domain.")
         self._basis_functions = basis_functions
 
+
+class MutableBasisFunctionSet(BasisFunctionSet):
+    """Set of basis functions where number of functions is mutable.
+
+    Parameters
+    ----------
+    num_terms : int
+        the number of terms in the set.
+    """
+
+    def __init__(self, num_terms):
+        self._num_terms = None
+        self._basis_functions = None
+
+        self.num_terms = num_terms
+
     @property
     def basis_functions(self):
         """list: Basis functions."""
+        if self._basis_functions is None:
+            self._create()
         return self._basis_functions
+
+    @property
+    def num_terms(self):
+        """int: the number of terms in the set."""
+        return self._num_terms
+
+    @num_terms.setter
+    def num_terms(self, value):
+        num_terms = int(value)
+        if num_terms <= 0:
+            raise ValueError("Must have at least one term.")
+        if num_terms != self._num_terms:
+            self._num_terms = num_terms
+            self._basis_functions = None
+
+    @abc.abstractmethod
+    def _create(self):
+        """Create the basis functions in the set."""
+        pass
+
+
+class ChebyshevFirstKindBasisFunctionSet(MutableBasisFunctionSet):
```

Comment:
Have this class derive directly from `BasisFunctionSet`, then give it an `__init__` method:
```
def __init__(self, num_terms):
    basis_functions = [ChebyshevFirstKind(n) for n in range(num_terms)]
    super().__init__(basis_functions)
```
Use a similar pattern for the two below.

The one "issue" with this approach is that all the basis functions are constructed when the set is, which is different from the point sets (which delay construction). I am OK with this if you are, since it is making objects (not data) immediately usable.